tasks/MessageDecoder.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef COMMS_WEBRTC_MESSAGE_DECODER_HPP
+#define COMMS_WEBRTC_MESSAGE_DECODER_HPP
+
+#include "json/json.h"
+#include "stdio.h"
+#include "string.h"
+#include "memory"
+
+namespace comms_webrtc
+{
+    struct MessageDecoder
+    {
+        Json::Value jdata;
+        Json::CharReaderBuilder rbuilder;
+        std::unique_ptr<Json::CharReader> const reader;
+
+        MessageDecoder() : reader(rbuilder.newCharReader()) {}
+
+        bool parseJSONMessage(char const *data, std::string &errors)
+        {
+            return reader->parse(data, data + strlen(data), &jdata, &errors);
+        }
+
+        void validateFieldPresent(std::string const &fieldName)
+        {
+            if (!jdata.isMember(fieldName))
+            {
+                throw std::invalid_argument(
+                    "message does not contain the " + fieldName + " field");
+            }
+        }
+
+        void validateDataFieldPresent(std::string const &fieldName)
```

Comment:
*This* is why you avoid passing state through an attribute.

Just change `validateFieldPresent` to have a `Json::Value&` argument as first argument and boom ... you only need one method.